### ISSUE TITLE
Workflow stuck in executing

### ISSUE THREAD

**Issue by Sinistr0** [Shuffle/Shuffle#979]
Hi, I have shuffle in a production environment, but in some workflows some events are stuck running for days, and they are extremely important events, they get stuck running for days. what can cause this and how to solve it?

follow images of my workflow:
![image](https://user-images.githubusercontent.com/82057628/200416388-0cd56588-bbf8-4191-a45e-bf79b43ea437.png)

![image](https://user-images.githubusercontent.com/82057628/200416443-39facbb4-a6f7-41e8-9799-bf74685999ef.png)

stuck here:

![image](https://user-images.githubusercontent.com/82057628/200416555-8197d227-92fb-4732-9662-d19cc81f5868.png)

**Comment by mvktsk** [comment]
Looks like we have the same issue. Sometimes some actions in workflow marks as skipped and workflow take long time to run (more 10 minutes), when workflow triggered by webhook. But when workflow runs manually, all actions run fast and workflow run takes less than a minute.

**Comment by frikky** [comment]
> Hi, I have shuffle in a production environment, but in some workflows some events are stuck running for days, and they are extremely important events, they get stuck running for days. what can cause this and how to solve it?
> 
> follow images of my workflow: ![image](https://user-images.githubusercontent.com/82057628/200416388-0cd56588-bbf8-4191-a45e-bf79b43ea437.png)
> 
> ![image](https://user-images.githubusercontent.com/82057628/200416443-39facbb4-a6f7-41e8-9799-bf74685999ef.png)
> 
> stuck here:
> 
> ![image](https://user-images.githubusercontent.com/82057628/200416555-8197d227-92fb-4732-9662-d19cc81f5868.png)

Hey! Could we set up a call and dig into this?

In most cases, we've found that the problem is too little resources on the "Orborus" server and that it's [not configured according to our recommended production readiness docs](https://shuffler.io/docs/configuration#production_readiness). The open source default setup of Shuffle is not built for production - more like a preview, and will need some tweeks to get to an ideal stage.

@mvktsk this goes for you too. Workflows should finish in a few seconds ideally - not minutes

**Comment by Sinistr0** [comment]
I'm using a vm with 16GB ram and 4v cpu, and shuffle is installed in a docker inside it, I think this is enough for the production environment, can you correct me if I'm wrong? what I find strange is the flows are running infinitely, I believe that a useful functionality would be after a certain time it restarts the execution process, because when I run these flows that are manually stuck they run successfully

**Comment by frikky** [comment]
> I'm using a vm with 16GB ram and 4v cpu, and shuffle is installed in a docker inside it, I think this is enough for the production environment, can you correct me if I'm wrong? what I find strange is the flows are running infinitely, I believe that a useful functionality would be after a certain time it restarts the execution process, because when I manually run these flows that are stuck they run successfully

A couple good points there. Further pointers:
- We recommend 2 servers for the single reason that CPU spikes for Orborus in the open source setup, which creates a chain effect if on the same server (orborus gets slow -> backend gets slow -> orborus doesn't get jobs... This is a cycle).
- Adding CPU/RAM is not the same as configuring the Backend & Orborus itself to be fast. Gotta utilize them efficiently as well
- There are execution restarts & abort systems in Shuffle. What version are you on? Max timeout for started workflows should be 30 minutes usually.

**Here are some options I suggest:**
- Set up multiple environments and split multiple Orborus containers across multiple servers  
- Work with us as a customer and get access to support and better scale licensing

**Comment by mvktsk** [comment]
@frikky we have split backend and Orborus.  Orborus dedicated server have 28 cores and 32 GB RAM  (docker deployment, configured as described in docs).
In our case, looks like some workflow branches did not send a signal to backend that workflow finished.
Are any possibilities to get workflow run log with job timings?
Something like [Azure Logic Apps history](https://learn.microsoft.com/en-us/azure/logic-apps/monitor-logic-apps?tabs=consumption) ![Azure Logic app run](https://learn.microsoft.com/en-us/azure/logic-apps/media/monitor-logic-apps/logic-app-run-pane-consumption.png)

**Comment by frikky** [comment]
> @frikky we have split backend and Orborus. Orborus dedicated server have 28 cores and 32 GB RAM (docker deployment, configured as described in docs). In our case, looks like some workflow branches did not send a signal to backend that workflow finished. Are any possibilities to get workflow run log with job timings? Something like [Azure Logic Apps history](https://learn.microsoft.com/en-us/azure/logic-apps/monitor-logic-apps?tabs=consumption) ![Azure Logic app run](https://camo.githubusercontent.com/f8bccffdfc3b4347f618445df6b9b867f58c3bc81258c8c0bddf56075780901b/68747470733a2f2f6c6561726e2e6d6963726f736f66742e636f6d2f656e2d75732f617a7572652f6c6f6769632d617070732f6d656469612f6d6f6e69746f722d6c6f6769632d617070732f6c6f6769632d6170702d72756e2d70616e652d636f6e73756d7074696f6e2e706e67)
> 
> PS We have only one workflow in Shuffle

In that case, I'm guessing it has more to do with concurrency problems than resource constraints. There are indeed more replay system that will be released in 1.1.0 among a lot of other things, and you can indeed dig into the executions as you're showing already (including reruns). Sounds like a server that is more than beefy enough to run this :)

**Comment by ssriley** [comment]
We experienced the same issue with one of our workflows as well. The workflow would hang if it was called from a parent workflow or if it was triggered by a webhook trigger. If we ran the workflow manually, it would run fine. The workflow would get stuck on various steps for no reason. I looked at the logs of the stuck worker and the only thing i could see that was maybe an issue was this:

Failed setting up Opensearch: dial tcp: lookup shuffle-opensearch on 127.0.0.11:53: server misbehaving

As a workaround, we started the workflow from the execute API and that seemed to work. Its not a fix but hopefully this helps someone in this situation.

**Comment by Sinistr0** [comment]
Has anyone managed to find a solution for this? I updated to the latest version and the same problem continued, and as mentioned above, it only happens in workflows triggered by webhooks, very strange.

**Comment by frikky** [comment]
> Has anyone managed to find a solution for this? I updated to the latest version and the same problem continued, and as mentioned above, it only happens in workflows triggered by webhooks, very strange.

Hey! We've created new Workers multiple times since back in December to handle these issues. Try restarting your Orborus again, and it may just end up solve the issue.

As for Opensearch, the issue is typically either of these:
- Too little RAM given to Opensearch in docker-compose.yml (we've set default to 2gb instead of 1gb nowadays)
- Not configured: $ sysctl -w vm.max_map_count=262144
